parser_library/test/processing/end_instr_test.cpp

```diff
@@ -0,0 +1,411 @@
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "A249" }));
+}
+TEST(END, three_operands_false)
+{
+    std::string input(R"( 
+NAME     CSECT
+AREA     DS              50F
+ENTRYPT  BALR            2,0
+     END ENTRYPT,SDS(MYCOMPILER,0101,00273),TEST
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "E010" }));
+}
+
+TEST(END, end_called_from_macro_correct)
+{
+    std::string input(R"( 
+         MACRO                                               
+     MAC                                                 
+     END   TEST                                   
+     MEND                                                
+TEST CSECT                                
+     MAC
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(a.diags().empty());
+}
+
+TEST(END, end_called_from_macro_with_statements_after_end)
+{
+    std::string input(R"( 
+         MACRO                                               
+     MAC                                                 
+     END   TEST                                         
+     MEND                                                
+TEST CSECT                                                                                 
+     MAC  
+     LR 1,2     
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "W015" }));
+}
+
+TEST(END, end_called_from_macro_with_lookahead_symbols_found)
+{
+    std::string input(R"( 
+         MACRO                                               
+     MAC                                                 
+     END   TEST                                          
+     MEND                                                
+TEST CSECT                                                   
+     AIF   (L'X LT 0).X                                 
+     MAC
+     BR    14                                            
+.X   ANOP  ,                                            
+X    DS    C                                                          
+     BR    0                                             
+     END                                                 
+)");
+    analyzer a(input);
+    a.analyze();
+    a.collect_diags();
+    EXPECT_TRUE(matches_message_codes(a.diags(), { "W015" }));
+}
+
+TEST(END, end_called_from_macro_inside_copy_book_one_with_warning_after_end)
+{
+    std::string copybook_content = R"(
+   MACRO
+  M
+  LR 1,2
+  END 
+ MEND       
+)";
+    std::string input = R"(
+         COPY COPYBOOK_TWO
+         M
+         LR 1,2
+)";
+    mock_parse_lib_provider lib_provider { { "COPYBOOK_TWO", copybook_content } };
+    analyzer a(input, analyzer_options { &lib_provider });
+
+    a.analyze();
+    a.collect_diags();
+    EXPECT_EQ(1, std::count_if(a.diags().begin(), a.diags().end(), [](const auto& d) { return d.code == "W015"; }));
```

Comment:
```suggestion
    EXPECT_TRUE(matches_message_codes(a.diags(), { "W015" }));
```
or is something else expected?